File: platform/sdk/swift/Sources/LayerXSDK/Stream.swift
import Foundation

public struct StreamCursor: RawRepresentable, Hashable, Sendable, Codable, CustomStringConvertible {
    public let rawValue: String

    public init?(rawValue: String) {
        guard !rawValue.isEmpty, rawValue.utf8.count <= 512, !rawValue.contains("\0") else { return nil }
        self.rawValue = rawValue
    }

    public init(_ value: String) throws {
        guard let cursor = Self(rawValue: value) else {
            throw PlatformSDKError(code: .invalidArgument, retry: .never)
        }
        self = cursor
    }

    public var description: String { rawValue }
}

public struct StreamEvent<Value: Sendable>: Sendable {
    public let eventID: String
    public let previousCursor: StreamCursor
    public let cursor: StreamCursor
    public let value: Value

    public init(eventID: String, previousCursor: StreamCursor, cursor: StreamCursor, value: Value) {
        self.eventID = eventID
        self.previousCursor = previousCursor
        self.cursor = cursor
        self.value = value
    }
}

public struct StreamPage<Value: Sendable>: Sendable {
    public let requestedCursor: StreamCursor
    public let events: [StreamEvent<Value>]
    public let nextCursor: StreamCursor

    public init(requestedCursor: StreamCursor, events: [StreamEvent<Value>], nextCursor: StreamCursor) {
        self.requestedCursor = requestedCursor
        self.events = events
        self.nextCursor = nextCursor
    }
}

public typealias StreamPageSource<Value: Sendable> = @Sendable (StreamCursor) async throws -> StreamPage<Value>

public actor ResumableStream<Value: Sendable> {
    public private(set) var cursor: StreamCursor
    private var seenEventIDs: Set<String> = []

    public init(cursor: StreamCursor) {
        self.cursor = cursor
    }

    public func accept(_ page: StreamPage<Value>) throws -> [StreamEvent<Value>] {
        guard page.requestedCursor == cursor else {
            throw PlatformSDKError(code: .decodeFailure, retry: .never)
        }
        var expected = cursor
        var pageEventIDs: Set<String> = []
        var accepted: [StreamEvent<Value>] = []
        accepted.reserveCapacity(page.events.count)
        for event in page.events {
            guard !event.eventID.isEmpty,
                  event.previousCursor == expected,
                  event.cursor != event.previousCursor,
                  !seenEventIDs.contains(event.eventID),
                  pageEventIDs.insert(event.eventID).inserted else {
                throw PlatformSDKError(code: .decodeFailure, retry: .never)
            }
            accepted.append(event)
            expected = event.cursor
        }
        guard page.nextCursor == expected else {
            throw PlatformSDKError(code: .decodeFailure, retry: .never)
        }

        seenEventIDs.formUnion(pageEventIDs)
        cursor = page.nextCursor
        return accepted
    }

    public func next(from source: StreamPageSource<Value>) async throws -> [StreamEvent<Value>] {
        let requested = cursor
        let page = try await source(requested)
        return try accept(page)
    }

    public nonisolated func events(from source: @escaping StreamPageSource<Value>) -> AsyncThrowingStream<StreamEvent<Value>, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    while !Task.isCancelled {
                        for event in try await self.next(from: source) {
                            continuation.yield(event)
                        }
                    }
                    continuation.finish()
                } catch is CancellationError {
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

public struct HumanStreamSource: Sendable {
    private let client: PlatformClient

    public init(client: PlatformClient) {
        self.client = client
    }

    public func open() async throws -> StreamCursor {
        let response = try await client.humanStreamOpen()
        return try StreamCursor(try stringField("cursor", in: response))
    }

    public func next(after requested: StreamCursor) async throws -> StreamPage<JSONValue> {
        let response = try await client.humanStreamNext(pathParameters: ["cursor": requested.rawValue])
        guard let object = response.objectValue,
              case let .array(untrustedEvents)? = object["events"],
              case let .string(nextValue)? = object["next_cursor"] else { throw decodeFailure() }
        var previous = requested
        var events: [StreamEvent<JSONValue>] = []
        events.reserveCapacity(untrustedEvents.count)
        for value in untrustedEvents {
            let cursor = try StreamCursor(try stringField("cursor", in: value))
            events.append(StreamEvent(eventID: cursor.rawValue, previousCursor: previous, cursor: cursor, value: value))
            previous = cursor
        }
        return StreamPage(requestedCursor: requested, events: events, nextCursor: try StreamCursor(nextValue))
    }

    public func source() -> StreamPageSource<JSONValue> {
        { cursor in try await next(after: cursor) }
    }

    private func stringField(_ name: String, in value: JSONValue) throws -> String {
        guard let object = value.objectValue, case let .string(field)? = object[name] else { throw decodeFailure() }
        return field
    }

    private func decodeFailure() -> PlatformSDKError {
        PlatformSDKError(code: .decodeFailure, retry: .never)
    }
}

File: agent/sdk/python/layerx_sdk/stream.py
from __future__ import annotations

from collections.abc import Callable, Iterator
from dataclasses import dataclass
from typing import Generic, TypeVar, cast

from .production import PlatformSdkError, SdkErrorCode

T = TypeVar("T")


class StreamCursor(str):
    def __new__(cls, value: str) -> StreamCursor:
        if not value or len(value) > 512 or "\0" in value:
            raise PlatformSdkError(SdkErrorCode.INVALID_ARGUMENT, "never")
        return cast(StreamCursor, str.__new__(cls, value))


@dataclass(frozen=True)
class StreamEvent(Generic[T]):
    event_id: str
    previous_cursor: StreamCursor
    cursor: StreamCursor
    value: T


@dataclass(frozen=True)
class StreamPage(Generic[T]):
    requested_cursor: StreamCursor
    events: tuple[StreamEvent[T], ...]
    next_cursor: StreamCursor


class ResumableStream(Generic[T]):
    def __init__(self, cursor: StreamCursor) -> None:
        self._cursor = cursor
        self._seen: set[str] = set()

    @property
    def cursor(self) -> StreamCursor:
        return self._cursor

    def accept(self, page: StreamPage[T]) -> tuple[StreamEvent[T], ...]:
        if page.requested_cursor != self._cursor:
            raise PlatformSdkError(SdkErrorCode.DECODE_FAILURE, "never")
        expected = self._cursor
        accepted: list[StreamEvent[T]] = []
        for event in page.events:
            if not event.event_id or event.previous_cursor != expected or event.event_id in self._seen:
                raise PlatformSdkError(SdkErrorCode.DECODE_FAILURE, "never")
            self._seen.add(event.event_id)
            accepted.append(event)
            expected = event.cursor
        if page.next_cursor != expected:
            raise PlatformSdkError(SdkErrorCode.DECODE_FAILURE, "never")
        self._cursor = page.next_cursor
        return tuple(accepted)

    def events(self, source: Callable[[StreamCursor], StreamPage[T]]) -> Iterator[StreamEvent[T]]:
        while True:
            yield from self.accept(source(self._cursor))
